### PR TITLE
[Test] Add e2e test for sample RayJob yaml on kind

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?

Adds a test for the `RayJob` sample YAML to the Github Actions CI, similar to the existing `RayCluster` tests.

There is now a lot of repeated code in the `RayJob`, `RayService` and `RayCluster` tests, we should refactor this in the future.

Currently the test doesn't use any Rules, it just waits for the RayJob status to be SUCCEEDED. (This status originates from Ray.)

Passed 10/10 times locally.  
## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
@kevin85421 If buildkite is ready, I can try to run these tests on Buildkite instead of Github actions!

File: tests/framework/prototype.py
Comment (kevin85421) [inline tests/framework/prototype.py:424]: Will RayJob automatically clean up the Ray Pods after it succeeds?

**Review by kevin85421 — COMMENTED** [review]
Thank you for this contribution! Do you enable `pylint` in your IDE?

**Comment by kevin85421** [comment]
The standard runner in free GitHub Actions plan only has two CPUs. Hence, the test may be failed due to insufficient CPU. In that case, you can specify CPU in resource request/limit for both head/worker Pods. See https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray-cluster.external-redis.yaml for more details. The YAML is tested in GitHub Actions at this moment.

**Comment by architkulkarni** [comment]
> Thank you for this contribution! Do you enable `pylint` in your IDE?

Ah sorry, didn't think about linting.  If `pylint` is used for this project, shall I add it to https://github.com/ray-project/kuberay/blob/master/CONTRIBUTING.md? I could also make an issue to add the python linter to CI

File: tests/framework/prototype.py
Comment (architkulkarni) [reply tests/framework/prototype.py:227]: Gotcha, thanks for reminding me the context.

File: tests/framework/prototype.py
Comment (architkulkarni) [reply tests/framework/prototype.py:222]: Yeah it's unused currently, let me improve it according to your comment above and add it back.

File: tests/framework/prototype.py
Comment (architkulkarni) [reply tests/framework/prototype.py:424]: By default, the terminate cluster on completion flag is set to False, though we may change this behavior in the future.  (I'm not sure it's intentional that the default is False, and it doesn't seem to be documented at the moment)

**Review by kevin85421 — COMMENTED** [review]
Would you mind running 10 times consecutively to check the flakiness of this test and add the result to the PR description? Thanks!

**Comment by kevin85421** [comment]
cc @Yicheng-Lu-llll for review

File: tests/framework/prototype.py
Comment (architkulkarni) [reply tests/framework/prototype.py:237]: Ah makes sense... let me think of a more programmatic approach, or at the very least match on a longer and more precise string

**Comment by architkulkarni** [comment]
When I run the test repeatedly (even just with `EasyJobRule()`), it fails pretty often with `"Failed to start Job Supervisor actor: The name _ray_internal_job_actor_rayjob-sample-zwtkn (namespace=SUPERVISOR_ACTOR_RAY_NAMESPACE) is already taken. Please use a different name or get the existing actor using ray.get_actor('_ray_internal_job_actor_rayjob-sample-zwtkn', namespace='SUPERVISOR_ACTOR_RAY_NAMESPACE')."`

It might be related to https://github.com/ray-project/ray/issues/31356, I'll try to prioritize and fix this one.  It might be more important for the KubeRay use case, since I remember we have some issue where the same job is submitted twice by KubeRay.

File: ray-operator/config/samples/ray_v1alpha1_rayjob.yaml
Comment (kevin85421) [reply ray-operator/config/samples/ray_v1alpha1_rayjob.yaml:20]: > In that case, you can specify CPU in resource request/limit for both head/worker Pods.

https://github.com/ray-project/kuberay/blob/f747a99d4269b337888dc5d1ec7833e9bf846e5d/ray-operator/config/samples/ray-cluster.external-redis.yaml#L153-L157

> Also, should I move it to Buildkite CI, or is that part not ready yet?

We should move the test to Buildkite CI, but we can merge this PR as soon as possible, and open a new PR to move it to Buildkite CI. Does it make sense?

Comment (architkulkarni) [reply ray-operator/config/samples/ray_v1alpha1_rayjob.yaml:20]: Yup, makes sense

**Comment by architkulkarni** [comment]
It failed 5/10 times with the above error.  It seems the `RayJob` feature itself is flaky, not the test code. I think I should fix https://github.com/ray-project/kuberay/issues/756 first (using your idea of running `ray job submit` in the container commands) before merging this PR, to avoid making CI flaky.

File: .github/workflows/actions/configuration/action.yaml
Comment (architkulkarni) [inline .github/workflows/actions/configuration/action.yaml:86]:
```suggestion
      OPERATOR_IMAGE: kuberay/operator:v0.5.0 # The operator image in the latest KubeRay release.
```

**Comment by architkulkarni** [comment]
The test passed 9/10 times locally.  The one failure was a failure of `kind create cluster`: `listen tcp4 127.0.0.1:52365: bind: address already in use.` which I assume is from me running the tests right after one another.

The test also passed in Github Actions CI.

**Comment by architkulkarni** [comment]
@kevin85421 my fault, forgot to enable the new test, no need to review it now. I'll let you know when it's ready for review

**Comment by architkulkarni** [comment]
Hi @kevin85421 , I think the PR is ready to be merged now after your next review.
- Passed 10/10 times locally
- Passed in Github Actions CI
- I removed the `RayJobSuccessRule`.  We can't actually get the status of the job from the custom resource using `ray job status`, because currently we internally append random letters to the sample job name.  If there's another way to get the status, we can add it later, or we can wait until after we refactor `RayJob` to be a k8s job.
- The "SUCCEEDED" status is checked in the `RayJobAddCREvent` itself.  (The event is not considered "converged" until the job has succeeded.)

**Review by kevin85421 — APPROVED** [review]
LGTM

**Pull request merged**
